torch/onnx/_backend/core.py

```diff
@@ -0,0 +1,852 @@
+        ORTC.OrtDevice(
+            _get_ort_device_type(value.device.type),
+            ORTC.OrtDevice.default_memory(),
+            _device_id_or_zero(value.device.index),
+        )
+        for value in values
+    )
+    return devices
+
+
+def _get_ortvalues_from_torch_tensors(
+    tensors: Tuple[torch.Tensor, ...], devices: Tuple[ORTC.OrtDevice, ...]
+) -> Tuple[torch.Tensor, ...]:
+    ortvalues = ORTC.OrtValueVector()
+    ortvalues.reserve(len(tensors))
+    dtypes = []
+    shapes = []
+    data_ptrs = []
+
+    for tensor in tensors:
+        dtypes.append(_NP_DTYPE[tensor.dtype])
+        shapes.append(tensor.size())
+        data_ptrs.append(tensor.data_ptr())
+    ortvalues.push_back_batch(tensors, data_ptrs, dtypes, shapes, devices)
+    return ortvalues
+
+
+def _to_real_tensor(tensor: FakeTensor) -> torch.Tensor:
+    if tensor.is_sparse:
+        raise ValueError("sparse tensor is not yet supported.")
+    out = torch.empty(tensor.size(), dtype=tensor.dtype, device=tensor.device)
+    return out
+
+
+def _run_onnx_session_with_ortvaluevector(
+    sess: onnxruntime.InferenceSession,
+    input_names: Tuple[str, ...],
+    inputs: Tuple[torch.Tensor, ...],
+    input_devices: Tuple[ORTC.OrtDevice, ...],
+    output_names: Tuple[str, ...],
+    outputs: Tuple[torch.Tensor, ...],
+    output_devices: Tuple[ORTC.OrtDevice, ...],
+    preallocate_output: bool,
+) -> Tuple[torch.Tensor, ...]:
+    _nvtx_range_push("contiguous")
+    inputs = tuple(a.contiguous() for a in inputs)
+    _nvtx_range_pop()
+
+    _nvtx_range_push("push_back_batch")
+
+    ort_inputs = _get_ortvalues_from_torch_tensors(inputs, input_devices)
+
+    # preallocate output pytorch Tensors and use the buffers affined to the torch device for the output ortvalue.
+    # Because the output ortvalue is not allocated and owned by ort, it does not need to convert the output ortvalue
+    # to torch Tensor transferring the ownership.
+    if preallocate_output:
+        pth_outputs = tuple(
+            _to_real_tensor(t) if isinstance(t, FakeTensor) else t for t in outputs
+        )
+        ort_outputs = _get_ortvalues_from_torch_tensors(pth_outputs, output_devices)
+    else:
+        ort_outputs = ORTC.OrtValueVector()
+    _nvtx_range_pop()
+
+    _nvtx_range_push("run_with_ortvaluevector")
+    run_options = onnxruntime.RunOptions()
+    run_options.add_run_config_entry("disable_synchronize_execution_providers", "1")
+    sess.run_with_ortvaluevector(
+        run_options, input_names, ort_inputs, output_names, ort_outputs, output_devices
+    )
+    _nvtx_range_pop()
+
+    if preallocate_output:
+        return pth_outputs
+    else:
+        _nvtx_range_push("after run_with_ortvaluevector")
+        pth_outputs = onnxruntime.training.ortmodule._utils._ortvalues_to_torch_tensor(
+            ort_outputs
+        )
+        _nvtx_range_pop()
+        return pth_outputs
+
+
+def _assert_allclose_with_detailed_error_message(
```

Comment:
Can Pytorch's [assertEqual](https://github.com/pytorch/pytorch/blob/main/torch/testing/_internal/common_utils.py#L3188) be used instead? maybe [torch.testing.assert_close](https://github.com/pytorch/pytorch/blob/main/torch/testing/_comparison.py#L1250)?

Reply:
Sure. Will use torch.testing.assert_close.